### PR TITLE
ReadStreamToArraySegment use BufferListStream directly when available

### PR DESCRIPTION
**Pull request by danielmarbach**
Additional split of https://github.com/Azure/azure-sdk-for-net/pull/19683

# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running generate.ps1. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version,

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs [hunk 1]
```diff
@@ -139,10 +139,17 @@ private static ArraySegment<byte> ReadStreamToArraySegment(Stream stream)
                 return new ArraySegment<byte>();
             }
 
-            using var memStream = new MemoryStream(StreamBufferSizeInBytes);
-            stream.CopyTo(memStream, StreamBufferSizeInBytes);
-
-            return new ArraySegment<byte>(memStream.ToArray());
+            switch (stream)
```
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs:142]:
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
AMD Ryzen 9 3950X, 1 CPU, 32 logical and 16 physical cores
.NET Core SDK=5.0.201
  [Host]     : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT
  Job-AYYPIA : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT

InvocationCount=800000  

```
|                 Method |     Mean |   Error |   StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |---------:|--------:|---------:|------:|--------:|-------:|------:|------:|----------:|
| ConvertViaMemoryStream | 306.6 ns | 6.13 ns | 15.94 ns |  1.00 |    0.00 | 0.1100 |     - |     - |     928 B |
|           PatternMatch | 132.8 ns | 1.13 ns |  0.88 ns |  0.48 |    0.04 | 0.0325 |     - |     - |     280 B |

Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs:142]:
```
    using System;
    using System.IO;
    using System.Text;
    using BenchmarkDotNet.Attributes;
    using BenchmarkDotNet.Configs;
    using BenchmarkDotNet.Diagnosers;
    using BenchmarkDotNet.Exporters;
    using BenchmarkDotNet.Jobs;
    using Microsoft.Azure.Amqp;

    [Config(typeof(Config))]
    public class BufferListStreamDirectly
    {
        private BufferListStream stream;
        private AmqpMessage amqpMessage;

        private class Config : ManualConfig
        {
            public Config()
            {
                AddExporter(MarkdownExporter.GitHub);
                AddDiagnoser(MemoryDiagnoser.Default);
                AddJob(Job.Default.WithInvocationCount(800000));
            }
        }

        [IterationSetup]
        public void Setup()
        {
            stream = new BufferListStream(new[] {new ArraySegment<byte>(Encoding.UTF8.GetBytes("Hello"))});
            amqpMessage = AmqpMessage.Create(stream, false);
        }

        [IterationCleanup]
        public void Cleanup()
        {
            amqpMessage.Dispose();
        }

        private const int StreamBufferSizeInBytes = 512;

        [Benchmark(Baseline = true)]
        public ArraySegment<byte> ConvertViaMemoryStream()
        {
            using var messageStream = amqpMessage.ToStream();
            using var memStream = new MemoryStream(StreamBufferSizeInBytes);
            messageStream.CopyTo(memStream, StreamBufferSizeInBytes);
            return new ArraySegment<byte>(memStream.ToArray());
        }

        [Benchmark]
        public ArraySegment<byte> PatternMatch()
        {
            using var messageStream = amqpMessage.ToStream();
            return messageStream switch
            {
                BufferListStream bufferListStream => bufferListStream.ReadBytes((int) stream.Length),
                _ => throw new InvalidOperationException()
            };
        }
    }
```

**Review by jsquire — APPROVED** [review]
Thanks, @danielmarbach!   I'll defer to Josh for approval and merge, but assuming that we're getting a `BufferListStream` back from the AMQP library in most cases, this looks like it could be a nice efficiency with little risk.

**Comment by JoshLove-msft** [comment]
> Thanks, @danielmarbach! I'll defer to Josh for approval and merge, but assuming that we're getting a `BufferListStream` back from the AMQP library in most cases, this looks like it could be a nice efficiency with little risk.

As discussed in the other PR, we do always get back a `BufferListStream`, but looking in [master](https://github.com/Azure/azure-amqp/blob/f9d7c10a6c77dfcc0189829c1f6a51abd8c45592/src/AmqpMessage.cs#L458) it has been changed to a `MemoryStream`, so this still makes me a bit nervous. I don't think there are imminent plans to ship out of master though, and if others think this is a good calculated risk I'm happy to go with it. Still, it would be awesome if we could add coverage for the default case via a unit test, but we can always file an issue for that.

**Comment by danielmarbach** [comment]
_sigh_ even the virtual was removed on that method. Basically, a complete breaking change also for all inheritors of that abstract class 🤷

**Pull request merged**